plugin/documents.py

```diff
@@ -606,46 +614,70 @@ def render_highlights_on_main_thread() -> None:
 
     # --- textDocument/complete ----------------------------------------------------------------------------------------
 
-    def _on_query_completions_async(self, resolve: ResolveCompletionsFn, location: int) -> None:
-        session = self.session('completionProvider', location)
-        if not session:
-            resolve([], 0)
+    def _on_query_completions_async(self, resolve_completion_list: ResolveCompletionsFn, location: int) -> None:
+        sessions = self.sessions('completionProvider')
+        if not sessions:
+            resolve_completion_list([], 0)
             return
         self.purge_changes_async()
-        can_resolve_completion_items = bool(session.get_capability('completionProvider.resolveProvider'))
-        config_name = session.config.name
-        session.send_request_async(
-            Request.complete(text_document_position_params(self.view, location), self.view),
-            lambda res: self._on_complete_result(res, resolve, can_resolve_completion_items, config_name),
-            lambda res: self._on_complete_error(res, resolve))
-
-    def _on_complete_result(self, response: Optional[Union[dict, List]], resolve: ResolveCompletionsFn,
-                            can_resolve_completion_items: bool, session_name: str) -> None:
-        response_items = []  # type: List[Dict]
-        flags = 0
+
+        completion_promises = []  # type: List[Promise[CompletionResponseWithSession]]
+        for session in sessions:
+
+            def completion_request() -> Promise:
+                return Promise(lambda resolve: session.send_request_async(
+                    Request.complete(text_document_position_params(self.view, location), self.view),
+                    lambda res: self._on_complete_result(res, resolve, session),
```

Comment:
Just a suggestion to remove `_on_complete_result` and just resolve here directly:

```suggestion
                    lambda res: resolve(res, session),
```

(although if you will follow my other comment then you'll have to pass more arguments).

Reply:
I like this :)

Reply:
https://github.com/sublimelsp/LSP/pull/1582/commits/14c0e41f02c480352ffed4d54dc6c1d6ad07b8d3